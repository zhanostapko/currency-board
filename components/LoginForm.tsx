"use client";

import { useActionState } from "react";
import SubmitButton from "./SubmitButton";
import { loginAction } from "@/app/actions/login";
import Input from "./Shared/Input";
import AuthFormWrapper from "./AuthWrapper";
import OTPForm from "./OTPForm";
import Link from "next/link";

const LoginForm = () => {
  const [state, formAction] = useActionState(loginAction, {
    error: null,
    step: "login",
    email: "",
    otp: null,
  });

  return (
    <AuthFormWrapper
      title="Hello Again!"
      subtitle={state.step === "otp" ? "" : "Welcome Back"}
    >
      {state.step === "otp" ? (
        <>
          <OTPForm otpCode={state.otp} email={state.email} />
        </>
      ) : (
        <>
          <form className="flex flex-col gap-3 mb-4" action={formAction}>
            <Input
              label="Email Address"
              type="email"
              name="email"
              required
              defaultValue={state.email}
            />
            <Input label="Password" type="password" name="password" required />
            {state.error && (
              <p className="text-red-500 text-xs">{state.error}</p>
            )}
            <SubmitButton title="Sign In" />
          </form>

          <Link
            href="/register"
            className="text-blue-500 mt-6 text-sm cursor-pointer hover:underline"
          >
            Don&apos;t have an account? Sign up
          </Link>
          <p className="text-gray-500 mt-2 text-sm cursor-pointer hover:underline">
            Forgot Password
          </p>
        </>
      )}
    </AuthFormWrapper>
  );
};

export default LoginForm;
