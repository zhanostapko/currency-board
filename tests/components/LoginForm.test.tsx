import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import LoginForm from "../../components/LoginForm";

jest.mock("../../components/OTPForm", () => {
  const MockOTPForm = () => <div data-testid="otp-form">OTP Form</div>;
  MockOTPForm.displayName = "MockOTPForm";
  return MockOTPForm;
});

test("Login form renders correctly", () => {
  const { asFragment } = render(<LoginForm />);
  expect(screen.getByLabelText(/email address/i)).toBeInTheDocument();
  expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /sign in/i })).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

test("Shows error message for incorrect email", async () => {
  const { asFragment } = render(<LoginForm />);

  const emailInput = screen.getByLabelText(/email address/i);
  const passwordInput = screen.getByLabelText(/password/i);
  const signInButton = screen.getByRole("button", { name: /sign in/i });

  fireEvent.change(emailInput, { target: { value: "incorrect@email.com" } });
  fireEvent.change(passwordInput, { target: { value: "password123" } });
  fireEvent.click(signInButton);

  await waitFor(() => {
    expect(screen.getByText("Invalid email or password")).toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});

test("Shows error message for incorrect password", async () => {
  const { asFragment } = render(<LoginForm />);

  const emailInput = screen.getByLabelText(/email address/i);
  const passwordInput = screen.getByLabelText(/password/i);
  const signInButton = screen.getByRole("button", { name: /sign in/i });

  fireEvent.change(emailInput, { target: { value: "test@example.com" } });
  fireEvent.change(passwordInput, { target: { value: "incorrect-password" } });
  fireEvent.click(signInButton);

  await waitFor(() => {
    expect(screen.getByText("Invalid email or password")).toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});

test("Shows error message for invalid email format", async () => {
  const { asFragment } = render(<LoginForm disableValidation={true} />);

  const emailInput = screen.getByLabelText(/email address/i);
  const passwordInput = screen.getByLabelText(/password/i);
  const signInButton = screen.getByRole("button", { name: /sign in/i });

  fireEvent.change(emailInput, { target: { value: "invalid-email" } });
  fireEvent.change(passwordInput, { target: { value: "password123" } });
  fireEvent.click(signInButton);

  await waitFor(() => {
    expect(screen.getByText("Invalid email format")).toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});
