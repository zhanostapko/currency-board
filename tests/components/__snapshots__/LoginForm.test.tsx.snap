// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Hello Again!
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Welcome Back
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-3 mb-4"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email Address
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign In
        </button>
      </form>
      <a
        class="text-blue-500 mt-6 text-sm cursor-pointer hover:underline"
        href="/register"
      >
        Don't have an account? Sign up
      </a>
      <p
        class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
      >
        Forgot Password
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Shows error message for incorrect email 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Hello Again!
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Welcome Back
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-3 mb-4"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email Address
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value="incorrect@email.com"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <p
          class="text-red-500 text-xs"
        >
          Invalid email or password
        </p>
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign In
        </button>
      </form>
      <a
        class="text-blue-500 mt-6 text-sm cursor-pointer hover:underline"
        href="/register"
      >
        Don't have an account? Sign up
      </a>
      <p
        class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
      >
        Forgot Password
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Shows error message for incorrect password 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Hello Again!
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Welcome Back
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-3 mb-4"
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email Address
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value="test@example.com"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <p
          class="text-red-500 text-xs"
        >
          Invalid email or password
        </p>
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign In
        </button>
      </form>
      <a
        class="text-blue-500 mt-6 text-sm cursor-pointer hover:underline"
        href="/register"
      >
        Don't have an account? Sign up
      </a>
      <p
        class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
      >
        Forgot Password
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Shows error message for invalid email format 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center "
  >
    <div
      class="bg-white p-10 rounded-xl shadow-xl w-96 text-center"
    >
      <h2
        class="text-3xl font-bold text-gray-900 mb-2"
      >
        Hello Again!
      </h2>
      <p
        class="text-gray-600 mb-6"
      >
        Welcome Back
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex flex-col gap-3 mb-4"
        novalidate=""
      >
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="email"
        >
          Email Address
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="email"
          name="email"
          required=""
          type="email"
          value="invalid-email"
        />
        <label
          class="text-gray-400 text-left text-sm font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="text-gray-600 w-full outline-none relative border rounded-lg flex items-center p-3  border-gray-300"
          id="password"
          name="password"
          required=""
          type="password"
        />
        <p
          class="text-red-500 text-xs"
        >
          Invalid email format
        </p>
        <button
          class="bg-blue-600 text-white py-3 rounded-lg text-center hover:bg-blue-700 transition"
          type="submit"
        >
          Sign In
        </button>
      </form>
      <a
        class="text-blue-500 mt-6 text-sm cursor-pointer hover:underline"
        href="/register"
      >
        Don't have an account? Sign up
      </a>
      <p
        class="text-gray-500 mt-2 text-sm cursor-pointer hover:underline"
      >
        Forgot Password
      </p>
    </div>
  </div>
</DocumentFragment>
`;
